src/io/internal/pipelines/dicom/apply-pstate-to-image.cxx

```diff
@@ -0,0 +1,582 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         https://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+/*
+ *
+ *  Copyright (C) 1998-2018, OFFIS e.V.
+ *  All rights reserved.  See COPYRIGHT file for details.
+ *
+ *  This software and supporting documentation were developed by
+ *
+ *    OFFIS e.V.
+ *    R&D Division Health
+ *    Escherweg 2
+ *    D-26121 Oldenburg, Germany
+ *
+ *
+ *  Module:  dcmpstat
+ *
+ *  Authors: Joerg Riesmeier, Marco Eichelberg
+ *
+ *  Purpose
+ *    sample application that reads a DICOM image and (optionally)
+ *    a presentation state and creates a PGM bitmap using the settings
+ *    of the presentation state. Non-grayscale transformations are
+ *    ignored. If no presentation state is loaded, a default is created.
+ *
+ */
+
+#include "itkPipeline.h"
+#include "itkOutputTextStream.h"
+#include "itkOutputBinaryStream.h"
+#include "itkOutputImage.h"
+
+#include "dcmtk/config/osconfig.h"    /* make sure OS specific configuration is included first */
+
+#include "dcmtk/ofstd/ofstream.h"
+// Fix warning for redefinition of __STDC_FORMAT_MACROS in the header include tree for dviface.h
+#ifdef __STDC_FORMAT_MACROS
+  #undef __STDC_FORMAT_MACROS
+#endif
+#include "dcmtk/dcmpstat/dviface.h"
+#include "dcmtk/dcmpstat/dvpstx.h"    /* for DVPSTextObject */
+#include "dcmtk/dcmpstat/dvpsgr.h"    /* for DVPSGraphicObject */
+#include "dcmtk/dcmpstat/dvpscu.h"    /* for DVPSCurve */
+#include "dcmtk/dcmimgle/dcmimage.h"
+#include "dcmtk/dcmdata/cmdlnarg.h"
+#include "dcmtk/ofstd/ofcmdln.h"
+#include "dcmtk/ofstd/ofconapp.h"
+#include "dcmtk/dcmdata/dcuid.h"      /* for dcmtk version name */
+#include "dcmtk/dcmdata/dcjson.h"     /* JSON format */
+
+#include "cpp-base64/base64.h"
+#include "rapidjson/document.h"
+#include "rapidjson/stringbuffer.h"
+#include "rapidjson/writer.h"
+#include "rapidjson/prettywriter.h"
+using namespace rapidjson;
+
+#include <cstring>
+#include <array>
+
+#ifdef WITH_ZLIB
+#include "itk_zlib.h"     /* for zlibVersion() */
+#endif
+
+#define OFFIS_CONSOLE_APPLICATION "dcmp2pgm"
+
+static OFLogger dcmp2pgmLogger = OFLog::getLogger("dcmtk.apps." OFFIS_CONSOLE_APPLICATION);
+
+static char rcsid[] = "$dcmtk: " OFFIS_CONSOLE_APPLICATION " v"
+  OFFIS_DCMTK_VERSION " " OFFIS_DCMTK_RELEASEDATE " $";
+
+template<typename Iteratorable>
+rapidjson::Value getArrayJson(Iteratorable container, rapidjson::Document::AllocatorType& allocator)
```

Comment:
Very nice! Should be generally useful. Perhaps move to itkPipeline.h inside the `itk::wasm::` namespace?

https://github.com/InsightSoftwareConsortium/itk-wasm/blob/a67d9f07a47ed4e99406f3f91d856a8c5610e9a8/include/itkPipeline.h#L106